pkg/etcdenvvar/envvarcontroller.go

```diff
@@ -128,15 +132,66 @@ func (c *EnvVarController) checkEnvVars() error {
 		return err
 	}
 
-	currEnvVarMap, err := getEtcdEnvVars(envVarContext{
+	ctx := envVarContext{
 		targetImagePullSpec:  c.targetImagePullSpec,
 		spec:                 *operatorSpec,
 		status:               *operatorStatus,
 		configmapLister:      c.configmapLister,
 		nodeLister:           c.nodeLister,
 		infrastructureLister: c.infrastructureLister,
 		networkLister:        c.networkLister,
-	})
+	}
+
+	bootstrapComplete, err := isBootstrapComplete(c.configmapLister, c.operatorClient)
+	if err != nil {
+		return fmt.Errorf("couldn't determine bootstrap status: %w", err)
+	}
+	isUnsupportedUnsafeEtcd, err := ceohelpers.IsUnsupportedUnsafeEtcd(&ctx.spec)
+	if err != nil {
+		return fmt.Errorf("couldn't determine etcd unsupported override status: %w", err)
+	}
+
+	// Detect semi-supported unsafe etcd bootstrapping by checking for an annotation
+	// on the etcd namespace.
+	isUnsafeBootstrapEnabled := false
+	etcdNamespace, err := c.namespaceLister.Get(operatorclient.TargetNamespace)
+	if err != nil {
+		return fmt.Errorf("couldn't get %s namespace: %w", operatorclient.TargetNamespace, err)
+	}
+	if _, hasAnnotation := etcdNamespace.Annotations[operatorclient.UnsafeBootstrapAnnotation]; hasAnnotation {
+		isUnsafeBootstrapEnabled = true
+	}
+
+	nodeCount := len(ctx.status.NodeStatuses)
+
+	// There are three discrete validation paths to enforce HA invariants
+	// depending on whether the unsupported/unsafe config is set, whether the
+	// cluster is managed by assisted installer, and the default configuration.
+	switch {
+	case isUnsupportedUnsafeEtcd:
+		if isUnsafeBootstrapEnabled {
+			return fmt.Errorf("unsupported/unsafe etcd mode and unsafe bootstrap mode are mutually exclusive")
+		}
```

Comment:
nit: the combined case could be the first case statement:
```
case isUnsupportedUnsafeEtcd && isUnsafeBootstrapEnabled:
```
But the current code is okay too.

Reply:
I'm not sure that's equivalent, but I could be wrong. There are three distinct operating "modes" here (unsupported, unsafe-bootstrap, and always-ha), and the first two are mutually exclusive. So I nested the mutual exclusivity check under the first so I could explicitly treat that as a distinct error. Hope that makes sense.

Reply:
They are equivalent, but what you have is fine.